// @medusajs/medusa/dist/services/payment-collection.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
var medusa_core_utils_1 = require("medusa-core-utils");
var utils_1 = require("../utils");
var models_1 = require("../models");
var interfaces_1 = require("../interfaces");
var PaymentCollectionService = /** @class */ (function (_super) {
    __extends(PaymentCollectionService, _super);
    function PaymentCollectionService(_a) {
        var manager = _a.manager, paymentCollectionRepository = _a.paymentCollectionRepository, paymentProviderService = _a.paymentProviderService, customerService = _a.customerService, eventBusService = _a.eventBusService;
        var _this = 
        // eslint-disable-next-line prefer-rest-params
        _super.call(this, arguments[0]) || this;
        _this.manager_ = manager;
        _this.paymentCollectionRepository_ = paymentCollectionRepository;
        _this.paymentProviderService_ = paymentProviderService;
        _this.eventBusService_ = eventBusService;
        _this.customerService_ = customerService;
        return _this;
    }
    PaymentCollectionService.prototype.retrieve = function (paymentCollectionId, config) {
        var _a;
        if (config === void 0) { config = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var manager, paymentCollectionRepository, query, paymentCollection;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        manager = (_a = this.transactionManager_) !== null && _a !== void 0 ? _a : this.manager_;
                        paymentCollectionRepository = manager.getCustomRepository(this.paymentCollectionRepository_);
                        query = (0, utils_1.buildQuery)({ id: paymentCollectionId }, config);
                        return [4 /*yield*/, paymentCollectionRepository.find(query)];
                    case 1:
                        paymentCollection = _b.sent();
                        if (!paymentCollection.length) {
                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_FOUND, "Payment collection with id ".concat(paymentCollectionId, " was not found"));
                        }
                        return [2 /*return*/, paymentCollection[0]];
                }
            });
        });
    };
    PaymentCollectionService.prototype.create = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                            var paymentCollectionRepository, paymentCollectionToCreate, paymentCollection;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        paymentCollectionRepository = manager.getCustomRepository(this.paymentCollectionRepository_);
                                        paymentCollectionToCreate = paymentCollectionRepository.create({
                                            region_id: data.region_id,
                                            type: data.type,
                                            status: models_1.PaymentCollectionStatus.NOT_PAID,
                                            currency_code: data.currency_code,
                                            amount: data.amount,
                                            metadata: data.metadata,
                                            created_by: data.created_by,
                                            description: data.description,
                                        });
                                        return [4 /*yield*/, paymentCollectionRepository.save(paymentCollectionToCreate)];
                                    case 1:
                                        paymentCollection = _a.sent();
                                        return [4 /*yield*/, this.eventBusService_
                                                .withTransaction(manager)
                                                .emit(PaymentCollectionService.Events.CREATED, paymentCollection)];
                                    case 2:
                                        _a.sent();
                                        return [2 /*return*/, paymentCollection];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentCollectionService.prototype.update = function (paymentCollectionId, data) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                            var paymentCollectionRepo, paymentCollection, _a, _b, key, result;
                            var e_1, _c;
                            return __generator(this, function (_d) {
                                switch (_d.label) {
                                    case 0:
                                        paymentCollectionRepo = manager.getCustomRepository(this.paymentCollectionRepository_);
                                        return [4 /*yield*/, this.retrieve(paymentCollectionId)];
                                    case 1:
                                        paymentCollection = _d.sent();
                                        try {
                                            for (_a = __values(Object.keys(data)), _b = _a.next(); !_b.done; _b = _a.next()) {
                                                key = _b.value;
                                                if (key === "metadata" && data.metadata) {
                                                    paymentCollection[key] = (0, utils_1.setMetadata)(paymentCollection, data.metadata);
                                                }
                                                else if ((0, utils_1.isDefined)(data[key])) {
                                                    paymentCollection[key] = data[key];
                                                }
                                            }
                                        }
                                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                                        finally {
                                            try {
                                                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                                            }
                                            finally { if (e_1) throw e_1.error; }
                                        }
                                        return [4 /*yield*/, paymentCollectionRepo.save(paymentCollection)];
                                    case 2:
                                        result = _d.sent();
                                        return [4 /*yield*/, this.eventBusService_
                                                .withTransaction(manager)
                                                .emit(PaymentCollectionService.Events.UPDATED, result)];
                                    case 3:
                                        _d.sent();
                                        return [2 /*return*/, result];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentCollectionService.prototype.delete = function (paymentCollectionId) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                            var paymentCollectionRepo, paymentCollection;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        paymentCollectionRepo = manager.getCustomRepository(this.paymentCollectionRepository_);
                                        return [4 /*yield*/, this.retrieve(paymentCollectionId).catch(function () { return void 0; })];
                                    case 1:
                                        paymentCollection = _a.sent();
                                        if (!paymentCollection) {
                                            return [2 /*return*/];
                                        }
                                        if ([
                                            models_1.PaymentCollectionStatus.CANCELED,
                                            models_1.PaymentCollectionStatus.NOT_PAID,
                                        ].includes(paymentCollection.status) === false) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_ALLOWED, "Cannot delete payment collection with status ".concat(paymentCollection.status));
                                        }
                                        return [4 /*yield*/, paymentCollectionRepo.remove(paymentCollection)];
                                    case 2:
                                        _a.sent();
                                        return [4 /*yield*/, this.eventBusService_
                                                .withTransaction(manager)
                                                .emit(PaymentCollectionService.Events.DELETED, paymentCollection)];
                                    case 3:
                                        _a.sent();
                                        return [2 /*return*/, paymentCollection];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentCollectionService.prototype.isValidTotalAmount = function (total, sessionsInput) {
        var sum = sessionsInput.reduce(function (cur, sess) { return cur + sess.amount; }, 0);
        return total === sum;
    };
    PaymentCollectionService.prototype.setPaymentSessions = function (paymentCollectionId, sessions) {
        return __awaiter(this, void 0, void 0, function () {
            var sessionsInput;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        sessionsInput = Array.isArray(sessions) ? sessions : [sessions];
                        return [4 /*yield*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                                var paymentCollectionRepository, payCol, customer, selectedSessionIds, paymentSessions, _loop_1, this_1, sessionsInput_1, sessionsInput_1_1, session, e_2_1, removeIds;
                                var e_2, _a;
                                var _b, _c;
                                return __generator(this, function (_d) {
                                    switch (_d.label) {
                                        case 0:
                                            paymentCollectionRepository = manager.getCustomRepository(this.paymentCollectionRepository_);
                                            return [4 /*yield*/, this.retrieve(paymentCollectionId, {
                                                    relations: ["region", "region.payment_providers", "payment_sessions"],
                                                })];
                                        case 1:
                                            payCol = _d.sent();
                                            if (payCol.status !== models_1.PaymentCollectionStatus.NOT_PAID) {
                                                throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_ALLOWED, "Cannot set payment sessions for a payment collection with status ".concat(payCol.status));
                                            }
                                            sessionsInput = sessionsInput.filter(function (session) {
                                                return !!payCol.region.payment_providers.find(function (_a) {
                                                    var id = _a.id;
                                                    return id === session.provider_id;
                                                });
                                            });
                                            if (!this.isValidTotalAmount(payCol.amount, sessionsInput)) {
                                                throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.UNEXPECTED_STATE, "The sum of sessions is not equal to ".concat(payCol.amount, " on Payment Collection"));
                                            }
                                            customer = undefined;
                                            selectedSessionIds = [];
                                            paymentSessions = [];
                                            _loop_1 = function (session) {
                                                var existingSession, inputData, paymentSession, paymentSession;
                                                return __generator(this, function (_e) {
                                                    switch (_e.label) {
                                                        case 0:
                                                            if (!!customer) return [3 /*break*/, 2];
                                                            return [4 /*yield*/, this_1.customerService_
                                                                    .withTransaction(manager)
                                                                    .retrieve(session.customer_id, {
                                                                    select: ["id", "email", "metadata"],
                                                                })];
                                                        case 1:
                                                            customer = _e.sent();
                                                            _e.label = 2;
                                                        case 2:
                                                            existingSession = (_b = payCol.payment_sessions) === null || _b === void 0 ? void 0 : _b.find(function (sess) { return session.session_id === (sess === null || sess === void 0 ? void 0 : sess.id); });
                                                            inputData = {
                                                                resource_id: payCol.id,
                                                                currency_code: payCol.currency_code,
                                                                amount: session.amount,
                                                                provider_id: session.provider_id,
                                                                customer: customer,
                                                                metadata: {
                                                                    resource_id: payCol.id,
                                                                },
                                                            };
                                                            if (!existingSession) return [3 /*break*/, 4];
                                                            return [4 /*yield*/, this_1.paymentProviderService_
                                                                    .withTransaction(manager)
                                                                    .updateSessionNew(existingSession, inputData)];
                                                        case 3:
                                                            paymentSession = _e.sent();
                                                            selectedSessionIds.push(existingSession.id);
                                                            paymentSessions.push(paymentSession);
                                                            return [3 /*break*/, 6];
                                                        case 4: return [4 /*yield*/, this_1.paymentProviderService_
                                                                .withTransaction(manager)
                                                                .createSessionNew(inputData)];
                                                        case 5:
                                                            paymentSession = _e.sent();
                                                            selectedSessionIds.push(paymentSession.id);
                                                            paymentSessions.push(paymentSession);
                                                            _e.label = 6;
                                                        case 6: return [2 /*return*/];
                                                    }
                                                });
                                            };
                                            this_1 = this;
                                            _d.label = 2;
                                        case 2:
                                            _d.trys.push([2, 7, 8, 9]);
                                            sessionsInput_1 = __values(sessionsInput), sessionsInput_1_1 = sessionsInput_1.next();
                                            _d.label = 3;
                                        case 3:
                                            if (!!sessionsInput_1_1.done) return [3 /*break*/, 6];
                                            session = sessionsInput_1_1.value;
                                            return [5 /*yield**/, _loop_1(session)];
                                        case 4:
                                            _d.sent();
                                            _d.label = 5;
                                        case 5:
                                            sessionsInput_1_1 = sessionsInput_1.next();
                                            return [3 /*break*/, 3];
                                        case 6: return [3 /*break*/, 9];
                                        case 7:
                                            e_2_1 = _d.sent();
                                            e_2 = { error: e_2_1 };
                                            return [3 /*break*/, 9];
                                        case 8:
                                            try {
                                                if (sessionsInput_1_1 && !sessionsInput_1_1.done && (_a = sessionsInput_1.return)) _a.call(sessionsInput_1);
                                            }
                                            finally { if (e_2) throw e_2.error; }
                                            return [7 /*endfinally*/];
                                        case 9:
                                            if (!((_c = payCol.payment_sessions) === null || _c === void 0 ? void 0 : _c.length)) return [3 /*break*/, 11];
                                            removeIds = payCol.payment_sessions
                                                .map(function (sess) { return sess.id; })
                                                .filter(function (id) { return !selectedSessionIds.includes(id); });
                                            if (!removeIds.length) return [3 /*break*/, 11];
                                            return [4 /*yield*/, paymentCollectionRepository.deleteMultiple(removeIds)];
                                        case 10:
                                            _d.sent();
                                            _d.label = 11;
                                        case 11:
                                            payCol.payment_sessions = paymentSessions;
                                            return [4 /*yield*/, paymentCollectionRepository.save(payCol)];
                                        case 12: return [2 /*return*/, _d.sent()];
                                    }
                                });
                            }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentCollectionService.prototype.refreshPaymentSession = function (paymentCollectionId, sessionId, sessionInput) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                            var paymentCollectionRepository, payCol, session, customer, inputData, sessionRefreshed;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        paymentCollectionRepository = manager.getCustomRepository(this.paymentCollectionRepository_);
                                        return [4 /*yield*/, paymentCollectionRepository.getPaymentCollectionIdBySessionId(sessionId, {
                                                relations: [
                                                    "region",
                                                    "region.payment_providers",
                                                    "payment_sessions",
                                                ],
                                            })];
                                    case 1:
                                        payCol = _a.sent();
                                        if (paymentCollectionId !== payCol.id) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_ALLOWED, "Payment Session ".concat(sessionId, " does not belong to Payment Collection ").concat(paymentCollectionId));
                                        }
                                        session = payCol.payment_sessions.find(function (sess) { return sessionId === (sess === null || sess === void 0 ? void 0 : sess.id); });
                                        if (!session) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_ALLOWED, "Session with id ".concat(sessionId, " was not found"));
                                        }
                                        return [4 /*yield*/, this.customerService_
                                                .withTransaction(manager)
                                                .retrieve(sessionInput.customer_id, {
                                                select: ["id", "email", "metadata"],
                                            })];
                                    case 2:
                                        customer = _a.sent();
                                        inputData = {
                                            resource_id: payCol.id,
                                            currency_code: payCol.currency_code,
                                            amount: session.amount,
                                            provider_id: session.provider_id,
                                            customer: customer,
                                        };
                                        return [4 /*yield*/, this.paymentProviderService_
                                                .withTransaction(manager)
                                                .refreshSessionNew(session, inputData)];
                                    case 3:
                                        sessionRefreshed = _a.sent();
                                        payCol.payment_sessions = payCol.payment_sessions.map(function (sess) {
                                            if (sess.id === sessionId) {
                                                return sessionRefreshed;
                                            }
                                            return sess;
                                        });
                                        if (session.payment_authorized_at && payCol.authorized_amount) {
                                            payCol.authorized_amount -= session.amount;
                                        }
                                        return [4 /*yield*/, paymentCollectionRepository.save(payCol)];
                                    case 4:
                                        _a.sent();
                                        return [2 /*return*/, sessionRefreshed];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentCollectionService.prototype.markAsAuthorized = function (paymentCollectionId) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                            var paymentCollectionRepo, paymentCollection, result;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        paymentCollectionRepo = manager.getCustomRepository(this.paymentCollectionRepository_);
                                        return [4 /*yield*/, this.retrieve(paymentCollectionId)];
                                    case 1:
                                        paymentCollection = _a.sent();
                                        paymentCollection.status = models_1.PaymentCollectionStatus.AUTHORIZED;
                                        paymentCollection.authorized_amount = paymentCollection.amount;
                                        return [4 /*yield*/, paymentCollectionRepo.save(paymentCollection)];
                                    case 2:
                                        result = _a.sent();
                                        return [4 /*yield*/, this.eventBusService_
                                                .withTransaction(manager)
                                                .emit(PaymentCollectionService.Events.PAYMENT_AUTHORIZED, result)];
                                    case 3:
                                        _a.sent();
                                        return [2 /*return*/, result];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentCollectionService.prototype.authorize = function (paymentCollectionId, context) {
        if (context === void 0) { context = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                            var paymentCollectionRepository, payCol, authorizedAmount, i, session, auth, inputData, _a, _b, payColCopy;
                            var _c;
                            return __generator(this, function (_d) {
                                switch (_d.label) {
                                    case 0:
                                        paymentCollectionRepository = manager.getCustomRepository(this.paymentCollectionRepository_);
                                        return [4 /*yield*/, this.retrieve(paymentCollectionId, {
                                                relations: ["payment_sessions", "payments"],
                                            })];
                                    case 1:
                                        payCol = _d.sent();
                                        if (payCol.authorized_amount === payCol.amount) {
                                            return [2 /*return*/, payCol];
                                        }
                                        if (!(payCol.amount <= 0)) return [3 /*break*/, 3];
                                        payCol.authorized_amount = 0;
                                        return [4 /*yield*/, paymentCollectionRepository.save(payCol)];
                                    case 2: return [2 /*return*/, _d.sent()];
                                    case 3:
                                        if (!((_c = payCol.payment_sessions) === null || _c === void 0 ? void 0 : _c.length)) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_ALLOWED, "You cannot complete a Payment Collection without a payment session.");
                                        }
                                        authorizedAmount = 0;
                                        i = 0;
                                        _d.label = 4;
                                    case 4:
                                        if (!(i < payCol.payment_sessions.length)) return [3 /*break*/, 8];
                                        session = payCol.payment_sessions[i];
                                        if (session.payment_authorized_at) {
                                            authorizedAmount += session.amount;
                                            return [3 /*break*/, 7];
                                        }
                                        return [4 /*yield*/, this.paymentProviderService_
                                                .withTransaction(manager)
                                                .authorizePayment(session, context)];
                                    case 5:
                                        auth = _d.sent();
                                        if (auth) {
                                            payCol.payment_sessions[i] = auth;
                                        }
                                        if (!((auth === null || auth === void 0 ? void 0 : auth.status) === models_1.PaymentSessionStatus.AUTHORIZED)) return [3 /*break*/, 7];
                                        authorizedAmount += session.amount;
                                        inputData = {
                                            amount: session.amount,
                                            currency_code: payCol.currency_code,
                                            provider_id: session.provider_id,
                                            resource_id: payCol.id,
                                            payment_session: auth,
                                        };
                                        _b = (_a = payCol.payments).push;
                                        return [4 /*yield*/, this.paymentProviderService_
                                                .withTransaction(manager)
                                                .createPaymentNew(inputData)];
                                    case 6:
                                        _b.apply(_a, [_d.sent()]);
                                        _d.label = 7;
                                    case 7:
                                        i++;
                                        return [3 /*break*/, 4];
                                    case 8:
                                        if (authorizedAmount === 0) {
                                            payCol.status = models_1.PaymentCollectionStatus.AWAITING;
                                        }
                                        else if (authorizedAmount < payCol.amount) {
                                            payCol.status = models_1.PaymentCollectionStatus.PARTIALLY_AUTHORIZED;
                                        }
                                        else if (authorizedAmount === payCol.amount) {
                                            payCol.status = models_1.PaymentCollectionStatus.AUTHORIZED;
                                        }
                                        payCol.authorized_amount = authorizedAmount;
                                        return [4 /*yield*/, paymentCollectionRepository.save(payCol)];
                                    case 9:
                                        payColCopy = _d.sent();
                                        return [4 /*yield*/, this.eventBusService_
                                                .withTransaction(manager)
                                                .emit(PaymentCollectionService.Events.PAYMENT_AUTHORIZED, payColCopy)];
                                    case 10:
                                        _d.sent();
                                        return [2 /*return*/, payCol];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentCollectionService.Events = {
        CREATED: "payment-collection.created",
        UPDATED: "payment-collection.updated",
        DELETED: "payment-collection.deleted",
        PAYMENT_AUTHORIZED: "payment-collection.payment_authorized",
    };
    return PaymentCollectionService;
}(interfaces_1.TransactionBaseService));
exports.default = PaymentCollectionService;
